File: app/src/main/java/n7/ad2/heroes/full/ResponsesSourceFactory.java
package n7.ad2.heroes.full;


import android.arch.paging.DataSource;

public class ResponsesSourceFactory extends DataSource.Factory<Integer, Response> {

    private final ResponsesStorage responsesStorage;
    private String search;
    private PositionalResponsesDataSource responsesDataSource;

    ResponsesSourceFactory(final ResponsesStorage responsesStorage, String search) {
        this.responsesStorage = responsesStorage;
        responsesStorage.setInvalidate(new Invalidate() {
            @Override
            public void invalidate() {
                if (responsesDataSource != null) {
                    responsesDataSource.invalidate();
                }
            }
        });
        this.search = search;
    }

    @Override
    public DataSource<Integer, Response> create() {
        responsesDataSource = new PositionalResponsesDataSource(responsesStorage, search);
        return responsesDataSource;
    }
}

File: app/src/main/java/n7/ad2/heroes/full/ResponsesSourceFactory.kt
package n7.ad2.heroes.full

import androidx.paging.DataSource
import n7.ad2.ui.heroInfo.domain.vo.VOResponse

class ResponsesSourceFactory(
        private val list: List<VOResponse>,
        private val search: String
) : DataSource.Factory<Int, VOResponse>() {

    override fun create(): DataSource<Int, VOResponse> {
        return PositionalResponsesDataSource(list, search)
    }

}